// napajs/lib/core/timers/timer-api.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
var timer_1 = require("./timer");
var binding = require('../../binding');
function setImmediate(func) {
    var args = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        args[_i - 1] = arguments[_i];
    }
    var timeout = new timer_1.Timeout(func, 0, 0, args);
    binding.setImmediate(timeout);
    return timeout;
}
exports.setImmediate = setImmediate;
function clearImmediate(immediate) {
    immediate._active = false;
}
exports.clearImmediate = clearImmediate;
function setTimeout(func, after) {
    var args = [];
    for (var _i = 2; _i < arguments.length; _i++) {
        args[_i - 2] = arguments[_i];
    }
    var timeout = new timer_1.Timeout(func, after, 0, args);
    binding.setTimers(timeout);
    return timeout;
}
exports.setTimeout = setTimeout;
function clearTimeout(timeout) {
    timeout._active = false;
}
exports.clearTimeout = clearTimeout;
function setInterval(func, after) {
    var args = [];
    for (var _i = 2; _i < arguments.length; _i++) {
        args[_i - 2] = arguments[_i];
    }
    var timeout = new timer_1.Timeout(func, after, after, args);
    binding.setTimers(timeout);
    return timeout;
}
exports.setInterval = setInterval;
function clearInterval(timeout) {
    timeout._active = false;
}
exports.clearInterval = clearInterval;
